controllers/vspheremachine_controller.go

```diff
@@ -105,41 +114,159 @@ func AddMachineControllerToManager(ctx *context.ControllerManagerContext, mgr ma
+			return err
+		}
+	}
 	return nil
 }
 
 type machineReconciler struct {
 	*context.ControllerContext
+	VMService       services.VSphereMachineService
+	networkProvider services.NetworkProvider
+	supervisorBased bool
 }
 
+// nolint:gocognit
 // Reconcile ensures the back-end state reflects the Kubernetes resource state intent.
 func (r machineReconciler) Reconcile(ctx goctx.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	if r.supervisorBased {
+		logger := r.Logger.WithName(req.Namespace).WithName(req.Name)
+		logger.V(3).Info("Starting Reconcile VSphereMachine")
+		// Fetch the VSphereMachine instance.
+		vsphereMachine := &vmwarev1.VSphereMachine{}
+		if err := r.Client.Get(r, req.NamespacedName, vsphereMachine); err != nil {
+			if apierrors.IsNotFound(err) {
+				return reconcile.Result{}, nil
+			}
+			return reconcile.Result{}, err
+		}
+
+		// Fetch the CAPI Machine.
+		machine, err := clusterutilv1.GetOwnerMachine(r, r.Client, vsphereMachine.ObjectMeta)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		var cluster *clusterv1.Cluster
+		if machine != nil {
+			cluster = r.fetchCAPICluster(machine, vsphereMachine)
+		}
+
+		// Fetch the VSphereCluster
+		vsphereCluster := &vmwarev1.VSphereCluster{}
+		var vsphereClusterErr error
+		if cluster != nil {
+			vsphereClusterName := client.ObjectKey{
+				Namespace: vsphereMachine.Namespace,
+				Name:      cluster.Spec.InfrastructureRef.Name,
+			}
+
+			vsphereClusterErr = r.Client.Get(r, vsphereClusterName, vsphereCluster)
+		}
+		// Create the patch helper.
+		patchHelper, err := patch.NewHelper(vsphereMachine, r.Client)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrapf(
+				err,
+				"failed to init patch helper for %s %s/%s",
+				vsphereMachine.GroupVersionKind(),
+				vsphereMachine.Namespace,
+				vsphereMachine.Name)
+		}
+		// Create the machine context for this request.
+		clusterContext := &vmware.ClusterContext{
+			ControllerContext: r.ControllerContext,
+			Cluster:           cluster,
+			VSphereCluster:    vsphereCluster,
+		}
+
+		machineContext := &vmware.SupervisorMachineContext{
+			BaseMachineContext: &context.BaseMachineContext{
+				ControllerContext: r.ControllerContext,
+				Cluster:           cluster,
+				Machine:           machine,
+				Logger:            r.Logger.WithName(req.Namespace).WithName(req.Name),
+			},
+			ClusterContext: clusterContext,
+			PatchHelper:    patchHelper,
+			VSphereCluster: vsphereCluster,
+			VSphereMachine: vsphereMachine,
+		}
+		defer func() {
+			r.Recorder.EmitEvent(vsphereMachine, "Reconcile", reterr, true)
+			if err := machineContext.Patch(); err != nil {
+				machineContext.Logger.Error(err, "patch failed", "machine", machineContext.String())
+				if reterr == nil {
+					reterr = err
+				}
+			}
+		}()
+
+		// Handle deleted machines
+		if !vsphereMachine.ObjectMeta.DeletionTimestamp.IsZero() {
+			return r.reconcileDelete(machineContext)
+		}
+
+		if vsphereClusterErr != nil {
```

Comment:
why are we checking the error here and not early on?

Reply:
Supervisor clusters don't wait for the machines to get removed before allowing the delete on themselves like VSphereClusters. So we only error out if it's not a delete on the machine i.e: reconcileNormal since that will need a cluster present.